contracts/RMM01Portfolio.sol

```diff
@@ -28,34 +29,58 @@ contract RMM01Portfolio is PortfolioVirtual {
      * movement.
      * @custom:reverts Underflows if current reserves of output token is less then next reserves.
      */
-    function _getLatestInvariantAndVirtualPrice(uint64 poolId)
+    function _getLatestInvariantAndVirtualPrice(
+        uint64 poolId,
+        bool sellAsset
+    )
         internal
         view
         returns (uint256 price, int256 invariant, uint256 updatedTau)
     {
         PortfolioPool storage pool = pools[poolId];
         updatedTau = pool.computeTau(block.timestamp);
 
-        (uint256 x, uint256 y) = pool.getVirtualPoolReservesPerLiquidityInWad();
+        uint256 reserveXPerLiquidity = pool.virtualX;
+        uint256 reserveYPerLiquidity = pool.virtualY;
+
+        // The reserve which sends tokens out in this swap must be rounded up
+        // to avoid the scenario that the remainder in a truncated output amount
+        // is not stolen in the swap.
+        if (sellAsset) {
+            // Swap X -> Y, Y is output, so round up Y.
+            reserveXPerLiquidity =
+                reserveXPerLiquidity.divWadDown(pool.liquidity);
+            reserveYPerLiquidity = reserveYPerLiquidity.divWadUp(pool.liquidity);
+        } else {
+            // Swap Y -> X, X is output, so round up X.
+            reserveXPerLiquidity = reserveXPerLiquidity.divWadUp(pool.liquidity);
+            reserveYPerLiquidity =
+                reserveYPerLiquidity.divWadDown(pool.liquidity);
+        }
+
         invariant = int128(
-            pool.invariantOf({R_x: x, R_y: y, timeRemainingSec: updatedTau})
+            pool.invariantOf({
+                R_x: reserveXPerLiquidity,
+                R_y: reserveYPerLiquidity,
+                timeRemainingSec: updatedTau
+            })
         );
 
         price = RMM01Lib.getPriceWithX({
-            R_x: x,
+            R_x: reserveXPerLiquidity,
             stk: pool.params.maxPrice,
             vol: pool.params.volatility,
             tau: updatedTau
         });
     }
 
     /// @inheritdoc Objective
-    function _beforeSwapEffects(uint64 poolId)
-        internal
-        override
-        returns (bool, int256)
-    {
-        (, int256 invariant,) = _getLatestInvariantAndVirtualPrice(poolId);
+    function _beforeSwapEffects(
+        uint64 poolId,
+        bool sellAsset
+    ) internal override returns (bool, int256) {
+        (, int256 invariant,) =
+            _getLatestInvariantAndVirtualPrice(poolId, sellAsset);
```

Comment:
Adds the parameter `sellAsset` to pass it into the updated `_getLatestInvariantAndVirtualPrice` function.